rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java

```diff
@@ -54,7 +58,13 @@
     private final RxClient.ServerInfo serverInfo;
     private final PoolConfig poolConfig;
     private final ScheduledExecutorService cleanupScheduler;
-    private final AtomicBoolean isShutdown = new AtomicBoolean();
+    private final AtomicInteger aquiredConnectionsCounter = new AtomicInteger();
+    //is needed to not accept any new connections that are aquired
+    private final AtomicBoolean isShutdownRequested = new AtomicBoolean();
+    //is needed to not perform the shutdown twice
+    private final AtomicBoolean isShutdownPerformed = new AtomicBoolean();
+    //is needed to not perform shutdown during the aquire of a connection
+    private final ReadWriteLock shutdownLock = new ReentrantReadWriteLock();
```

Comment:
> > There could always be already checked out connections which aren't too different from concurrent acquire and shutdown.

Shouldn't it be that in that case the aquiredConnectionsCounter not equal to zero and the shutdown would not take place then?

> This lock means that the shutdown can not proceed while acquires are happening. If the shutdown happens from the eventloop (it can for rejection of hosts), it blocks the eventloop.


I see your point. Blocking in the event loop is not a good idea.
> release should cleanup when aquiredConnectionsCounter is 0 and the pool is shutdown.
That is what it is doing in my implementation or did I miss something?

> The above approach will have some connections, when release happens after shutdown but between acquire checking for shutdown and actually acquiring, that will be checkout even when the pool is cleaned up. I think we are ok with such behaviors if we check for shutdown within onSubscribe

I somehow do not like the idea that there is still the possibility of a connection leak. I know that this will be a rare case but it can still happen. What about the following approach:
Instead of blocking in the ```shutdown``` method we just do nothing if we can't acquire the lock. If we can't acquire the lock it basically means that someone is just about to acquire a connection. That connection will at some time be returned and the connection pool is closed.
What could still happen is that acquiring the connection fails. In that case the connection would not be returned and the pool would not be closed. So if the acquire fails we check if we have to shut down the pool. This way there is no chance to return a connection from an already closed connection pool.